README.md
[![CircleCI](https://circleci.com/gh/Ben-Hu/metabasic.svg?style=svg)](https://circleci.com/gh/Ben-Hu/metabasic)
[![Actions](https://github.com/Ben-Hu/metabasic/workflows/ci/badge.svg)](https://github.com/Ben-Hu/metabasic/actions)
[![codecov](https://codecov.io/gh/Ben-Hu/metabasic/branch/master/graph/badge.svg)](https://codecov.io/gh/Ben-Hu/metabasic)
[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/Ben-Hu/metabasic.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/Ben-Hu/metabasic/context:python)
[![License](https://img.shields.io/github/license/Ben-Hu/metabasic)](https://github.com/Ben-Hu/metabasic/blob/master/LICENSE)
[![Tag](https://img.shields.io/github/v/tag/Ben-Hu/metabasic)](https://github.com/Ben-Hu/metabasic/releases)
[![PyPI](https://img.shields.io/pypi/v/metabasic?color=blue)](https://pypi.org/project/metabasic/)


# Metabasic
Dead simple client for interacting with the Metabase dataset API

## Install
```sh
pip install metabasic
```

## Examples
```python
from metabasic import Metabasic
domain = "https://my-metabase-domain.com"

# Authentication with an existing session
db = Metabasic(domain, session_id="foo", database_id=1)
db.query("SELECT * FROM bar")
db.get_dataframe("SELECT * FROM bar")

# Email/Password authentication
ga = Metabasic(domain, database_id=2).authenticate("foo@email.com", "password")
ga_query = {
    "ids": "ga:1234567890",
    "start-date": "30daysAgo",
    "end-date": "today",
    "metrics": "ga:someMetric",
    "dimensions": "ga:someDimension",
    "sort": "ga:someDimension",
    "max-results": 10000
}
ga.query(json.dumps(ga_query))

# Select a database interactively
m = (
  Metabasic(domain)
  .authenticate("foo@email.com", "password")
  .select_database()
)
```


CHANGELOG.md
# 0.5.0 (2020-06-30)
### Enhancements
  - Adding `raw_query(query, export_format)` to retrieve responses in one of the supported export formats for `/api/dataset/{export_format}` (`csv`, `json`, `xslx`)

### Bug fix
  - `get_dataframe` uses `raw_query` with `export_format="csv"` to guarantee that the result set has a column ordering equal to the ordering specified by the query statement

# 0.4.2 (2020-06-24)
### Enhancements
  - Adding use of api/dataset/json to circumvent metabase row limit of 2000

# 0.4.0 (2020-05-11)
### Enhancements
  - Adds get_dataframe utility method, returning a pandas DataFrame
  - Updates query method to return the raw 'data' of the API resposne


# 0.3.0 (2020-05-10)
### Enhancements
  - Update authenticate & select_database methods to allow chaining


# 0.2.0 (2020-05-09)
### Enhancements
  - Add better handling of unconfigured/unauthenticated clients


# 0.1.0 (2020-05-09)
### Enhancements
  - Initial version

tests/metabasic_test.py
import io
from unittest.mock import Mock

import pandas as pd
import pytest
from requests import Response

from metabasic import Metabasic
from metabasic.exceptions import AuthError, ConfigError


class TestQuery:
    @pytest.fixture
    def metabasic(self):
        return Metabasic("domain", session_id="123abc", database_id=123)

    def test_query_success(self, mocker, metabasic):
        json = [{"a": 1}, {"b": 2}, {"c": 3}]

        mock_response = Mock(Response, json=lambda: json)
        mock_response.status_code = 202
        mocker.patch("requests.post", return_value=mock_response)

        assert metabasic.query("SELECT * FROM tests") == json

    def test_query_error(self, mocker, metabasic):
        mock_response = Mock(Response)
        mock_response.status_code = 401
        mocker.patch("requests.post", return_value=mock_response)

        with pytest.raises(Exception, match=str(mock_response)):
            metabasic.query("SELECT * FROM tests")

    def test_unconfigured(self, mocker):
        metabasic = Metabasic("domain", session_id="123abc")

        with pytest.raises(Exception, match=str(ConfigError())):
            metabasic.query("SELECT * FROM tests")

    def test_unauthenticated(self, mocker):
        metabasic = Metabasic("domain", database_id=123)

        with pytest.raises(Exception, match=str(AuthError())):
            metabasic.query("SELECT * FROM tests")


class TestGetDataFrame:
    @pytest.fixture
    def metabasic(self):
        return Metabasic("domain", session_id="123abc", database_id=123)

    def test_get_dataframe_success(self, mocker, metabasic):
        csv = b"a,b,c\n1,2,3\n4,5,6"

        mock_response = Mock(Response, content=csv)
        mock_response.status_code = 202

        mocker.patch("requests.post", return_value=mock_response)
        expected = pd.read_csv(io.BytesIO(csv))

        result = metabasic.get_dataframe("SELECT * FROM tests")
        assert isinstance(result, pd.DataFrame)
        assert result.to_dict() == expected.to_dict()

    def test_get_dataframe_error(self, mocker, metabasic):
        mock_response = Mock(Response)
        mock_response.status_code = 401
        mocker.patch("requests.post", return_value=mock_response)

        with pytest.raises(Exception, match=str(mock_response)):
            metabasic.get_dataframe("SELECT * FROM tests")

    def test_unconfigured(self, mocker):
        metabasic = Metabasic("domain", session_id="123abc")

        with pytest.raises(Exception, match=str(ConfigError())):
            metabasic.get_dataframe("SELECT * FROM tests")

    def test_unauthenticated(self, mocker):
        metabasic = Metabasic("domain", database_id=123)

        with pytest.raises(Exception, match=str(AuthError())):
            metabasic.get_dataframe("SELECT * FROM tests")


class TestAuthenticate:
    @pytest.fixture
    def metabasic(self):
        return Metabasic("domain")

    def test_authenticate_success(self, mocker, metabasic):
        mock_response = Mock(Response, json=lambda: {"id": "abc123"})
        mock_response.status_code = 200
        mocker.patch("requests.post", return_value=mock_response)

        result = metabasic.authenticate("email", "password")
        assert isinstance(result, Metabasic)
        assert metabasic.session_id == "abc123"

    def test_authenticate_error(self, mocker, metabasic):
        mock_response = Mock(Response)
        mock_response.status_code = 401
        mocker.patch("requests.post", return_value=mock_response)

        with pytest.raises(Exception, match=str(mock_response)):
            metabasic.authenticate("email", "password")


class TestSelectDatabase:
    @pytest.fixture
    def metabasic(self):
        return Metabasic("domain", session_id="123abc", database_id=123)

    def test_select_database_success(self, mocker, metabasic):
        databases = [{"id": 1, "name": "foo"}, {"id": 2, "name": "bar"}]

        mock_response = Mock(Response, json=lambda: databases)
        mock_response.status_code = 200
        mocker.patch("requests.get", return_value=mock_response)

        mocker.patch("inquirer.prompt", return_value={"name": "foo"})
        result = metabasic.select_database()

        assert isinstance(result, Metabasic)
        assert metabasic.database_id == 1

    def test_select_database_error(self, mocker, metabasic):
        mock_response = Mock(Response)
        mock_response.status_code = 401
        mocker.patch("requests.get", return_value=mock_response)

        with pytest.raises(Exception, match=str(mock_response)):
            metabasic.select_database()

    def test_unauthenticated(self, mocker):
        metabasic = Metabasic("domain", database_id=123)

        with pytest.raises(Exception, match=str(AuthError())):
            metabasic.select_database()


metabasic/metabasic.py
import io
import json
import urllib.parse
from typing import Any, Dict, Optional

import inquirer
import pandas as pd
import requests

from .exceptions import AuthError, ConfigError


class Metabasic(object):
    """Instantiates a new Metabasic Client

    Arguments:
        domain (str): The domain of the metabase host (i.e. http://my-metabase-host.com)
        session_id (Optional[str]): The metabase session id.
        database_id (Optional[int]): The metabase database id.
    """

    def __init__(
        self,
        domain: str,
        session_id: Optional[str] = None,
        database_id: Optional[int] = None,
    ):
        self.domain: str = domain
        self.session_id: Optional[str] = session_id
        self.database_id: Optional[int] = database_id

    def query(self, query: str) -> Dict[Any, Any]:
        """Queries the currently selected database.

        Arguments:
            query (str): The query to run against the currently selected database.

        Raises:
            AuthError: Raised if the client is unauthenticated.
            ConfigError: Raised if the client is unconfigured.
            Exception: Raised if an unexpected response is received.

        Returns:
            (List[List[Any]]): The results of the query.
        """
        resp = self.raw_query(query, export_format="json")
        return resp.json()

    def raw_query(self, query: str, export_format: str) -> requests.Response:
        """Queries the currently selected database whose results should be in
        the specified export format.

        Arguments:
            query (str): The query to run against the currently selected database.
            export_format (str): Use one of "json", "csv", "xlsx" to retrieve the result
                                 dataset in the specified format.

        Raises:
            AuthError: Raised if the client is unauthenticated.
            ConfigError: Raised if the client is unconfigured.
            Exception: Raised if an unexpected response is received.

        Returns:
            requests.Response: The response of the /api/dataset/{export_format} request.
        """
        self.__check_auth()
        self.__check_config()

        headers = {
            "Content-Type": "application/x-www-form-urlencoded",
            "X-Metabase-Session": self.session_id,
        }

        body: Dict[str, Any] = {
            "database": self.database_id,
            "native": {"query": query},
            "type": "native",
        }

        data = "query=" + urllib.parse.quote(json.dumps(body))

        resp: requests.Response = requests.post(
            f"{self.domain}/api/dataset/{export_format}", data=data, headers=headers
        )

        if resp.status_code != 202:
            raise Exception(resp)

        return resp

    def get_dataframe(self, query: str) -> pd.DataFrame:
        """Queries the currently selected database, returning a Pandas DataFrame.

        Arguments:
            query (str): The query to run against the currently selected database.

        Returns:
            pd.DataFrame: The results of the query wrapped into a Pandas DataFrame.
        """
        # Request csv export format so that don't lose the column order as specified
        # in the query. The json format does not always return the results in order,
        # and simple '/api/dataset/' requests do not return more than 2000 rows.
        res = self.raw_query(query, export_format="csv")
        buffer = io.BytesIO(res.content)
        df = pd.read_csv(buffer)
        return df

    def authenticate(self, email: str, password: str) -> "Metabasic":
        """Authenticates the client instance with the given email & password.

        Arguments:
            email (str): The email address to authenticate with.
            password (str): The password to authenticate with.

        Raises:
            Exception: Raised if an unexpected response is received.
        """
        headers = {"Content-Type": "application/json"}
        body = {"username": email, "password": password}
        resp = requests.post(f"{self.domain}/api/session", json=body, headers=headers)

        if resp.status_code != 200:
            raise Exception(resp)

        self.session_id = resp.json()["id"]

        return self

    def select_database(self) -> "Metabasic":
        """Start an interactive selection of available databases

        Raises:
            AuthError: Raised if the client is unauthenticated.
            Exception: Raised if an unexpected response is received.
        """
        self.__check_auth()

        headers = {
            "Content-Type": "application/json",
            "X-Metabase-Session": self.session_id,
        }

        resp = requests.get(f"{self.domain}/api/database", headers=headers)

        if resp.status_code != 200:
            raise Exception(resp)

        databases = {db["name"]: db["id"] for db in resp.json()}

        prompt = [inquirer.List("name", message="Database", choices=databases.keys())]
        target = inquirer.prompt(prompt)

        self.database_id = databases[target["name"]]

        return self

    def __check_auth(self):
        if not self.session_id:
            raise AuthError(
                "Client is unauthenticated.",
                "Provide a session_id or start a new session with the authenticate method.",
            )

    def __check_config(self):
        if not self.database_id:
            raise ConfigError(
                "Client is not configured",
                "Provide a database_id or select a one with the select_database method.",
            )
